### PR TITLE
proposal for adding a few new methods

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**📝 Description**
proposal to add a few methods to get the results of GetValues similar to what was added for GetSolutionStepValue.
The important comment here is that only one of the 4 methods added here is strictly different to what can be already found in the code (in that one can retrieve only x,y instead of x,y,z if he wants to), however i am adding the complete list of setters getters for consistency with the SolutionStepValuesVector functions.

Please mark the PR with appropriate tags: 
- Feature

**🆕 Changelog**
proposing the addition of a few methods to the variable_utils

### CODE REVIEW

File: kratos/python/add_variable_utils_to_python.cpp [hunk 1]
```diff
@@ -471,6 +471,21 @@ void AddVariableUtilsToPython(pybind11::module &m)
                             const Variable<double>&,
                             const Vector&,
                             const unsigned int>(&VariableUtils::SetSolutionStepValuesVector))
+        .def("GetValuesVector", py::overload_cast<
+                            const ModelPart::NodesContainerType&,
+                            const Variable<array_1d<double,3>>&,
+                            const unsigned int>(&VariableUtils::GetValuesVector))
```
Comment (RiccardoRossi) [inline kratos/python/add_variable_utils_to_python.cpp:477]: of the 4, this is the only method that is strictly different from others. 
I am adding the 4 with consistency with the methods just above

File: kratos/utilities/variable_utils.cpp [hunk 2]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
         );
 }
 
+
+
+
+
+
+
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
```
Many spaces!

File: kratos/utilities/variable_utils.cpp [hunk 3]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
...
+KRATOS_API(KRATOS_CORE) Vector VariableUtils::GetValuesVector(
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp:533]: Why KRATOS_API?, with defining in the class declaration is enough

File: kratos/utilities/variable_utils.cpp [hunk 4]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
...
+                            const unsigned int Dimension
+                            )
+{
+    Vector out(rNodes.size()*Dimension);
+
+    IndexPartition<unsigned int>(rNodes.size()).for_each(
+        [&](unsigned int i){
+            auto& v = (rNodes.begin()+i)->GetValue(rVariable);
+            for(unsigned int k=0; k<Dimension; k++)
+                out[i*Dimension+k] = v[k];
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
            const auto& r_v = (rNodes.begin()+i)->GetValue(rVariable);
            for(unsigned int k=0; k<Dimension; k++)
                out[i*Dimension+k] = r_v[k];
```

File: kratos/utilities/variable_utils.cpp [hunk 5]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
...
+{
+    KRATOS_ERROR_IF(rData.size()%rNodes.size()!=0) << "Incompatible number of nodes and position data" << std::endl;
+
+    const unsigned int Dimension = rData.size()/rNodes.size();
+
+    IndexPartition<unsigned int>(rNodes.size()).for_each(
+        [&](unsigned int i){
+            auto& v = (rNodes.begin()+i)->GetValue(rVar);
+            for(unsigned int k=0; k<Dimension; k++)
+                v[k] = rData[i*Dimension+k];
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
            auto& r_v = (rNodes.begin()+i)->GetValue(rVar);
            for(unsigned int k=0; k<Dimension; k++)
                r_v[k] = rData[i*Dimension+k];
```

File: kratos/utilities/variable_utils.cpp [hunk 6]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
...
+            }
+        );
+    return out;
+}
+
+KRATOS_API(KRATOS_CORE) void VariableUtils::SetValuesVector(
+                            ModelPart::NodesContainerType& rNodes,
+                            const Variable<array_1d<double,3>>& rVar,
+                            const Vector& rData
+                            )
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
    ModelPart::NodesContainerType& rNodes,
    const Variable<array_1d<double,3>>& rVar,
    const Vector& rData
    )
```

File: kratos/utilities/variable_utils.cpp [hunk 7]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
...
+                out[i*Dimension+k] = v[k];
+            }
+        );
+    return out;
+}
+
+KRATOS_API(KRATOS_CORE) Vector VariableUtils::GetValuesVector(
+                            const ModelPart::NodesContainerType& rNodes,
+                            const Variable<double>& rVar
+                            )
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
    const ModelPart::NodesContainerType& rNodes,
    const Variable<double>& rVar
    )
```

File: kratos/utilities/variable_utils.cpp [hunk 8]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
...
+
+
+
+
+
+KRATOS_API(KRATOS_CORE) Vector VariableUtils::GetValuesVector(
+                            const ModelPart::NodesContainerType& rNodes,
+                            const Variable<array_1d<double,3>>& rVariable,
+                            const unsigned int Dimension
+                            )
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
    const ModelPart::NodesContainerType& rNodes,
    const Variable<array_1d<double,3>>& rVariable,
    const unsigned int Dimension
    )
```

File: kratos/utilities/variable_utils.cpp [hunk 9]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
...
+                v[k] = rData[i*Dimension+k];
+            }
+        );
+}
+
+KRATOS_API(KRATOS_CORE) void VariableUtils::SetValuesVector(
+                            ModelPart::NodesContainerType& rNodes,
+                            const Variable<double>& rVar,
+                            const Vector& rData
+                            )
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
    ModelPart::NodesContainerType& rNodes,
    const Variable<double>& rVar,
    const Vector& rData
    )
```

File: kratos/utilities/variable_utils.cpp [hunk 10]
```diff
@@ -530,6 +530,81 @@ KRATOS_API(KRATOS_CORE) void VariableUtils::SetSolutionStepValuesVector(
...
+{
+    KRATOS_ERROR_IF(rData.size()%rNodes.size()!=0) << "Incompatible number of nodes and position data" << std::endl;
+
+    IndexPartition<unsigned int>(rNodes.size()).for_each(
+        [&](unsigned int i){
+            auto& v = (rNodes.begin()+i)->GetValue(rVar);
+            v = rData[i];
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.cpp]:
```suggestion
            auto& r_v = (rNodes.begin()+i)->GetValue(rVar);
            r_v = rData[i];
```

File: kratos/utilities/variable_utils.h [hunk 11]
```diff
@@ -1556,6 +1556,71 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
                                 const unsigned int Step
                                 );
 
+/**
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
    /**
```

File: kratos/utilities/variable_utils.h [hunk 12]
```diff
@@ -1556,6 +1556,71 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
+     * @param rNodes array of nodes from which data will be extracted
+     * @param rVar the variable being addressed
+     * @param Step step in the database
+     * @param Dimension number of components in output
+     */
+    Vector GetValuesVector(
+                                const ModelPart::NodesContainerType& rNodes,
+                                const Variable<array_1d<double,3>>& rVar,
+                                const unsigned int Dimension=3
+                                );
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
        const ModelPart::NodesContainerType& rNodes,
        const Variable<array_1d<double,3>>& rVar,
        const unsigned int Dimension=3
        );
```

File: kratos/utilities/variable_utils.h [hunk 13]
```diff
@@ -1556,6 +1556,71 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
+     * and for Dimension==3 the 3 components at once
+     * also note that this version accesses NON HISTORICAL data
+     * @param rNodes array of nodes from which data will be extracted
+     * @param rVar the variable being addressed
+     * @param Step step in the database
+     */
+    Vector GetValuesVector(
+                                const ModelPart::NodesContainerType& rNodes,
+                                const Variable<double>& rVar
+                    );
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
    Vector GetValuesVector(
        const ModelPart::NodesContainerType& rNodes,
        const Variable<double>& rVar
        );
```

File: kratos/utilities/variable_utils.h [hunk 14]
```diff
@@ -1556,6 +1556,71 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
+     * @param rNodes array of nodes from which data will be extracted
+     * @param rVar the variable being addressed
+     * @param rData input vector (must be of size rNodes.size()*Dimension)
+     * @param Step database step to which we will write
+     */
+    void SetValuesVector(
+                                ModelPart::NodesContainerType& rNodes,
+                                const Variable<array_1d<double,3>>& rVar,
+                                const Vector& rData
+                                );
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
        ModelPart::NodesContainerType& rNodes,
        const Variable<array_1d<double,3>>& rVar,
        const Vector& rData
        );
```

File: kratos/utilities/variable_utils.h [hunk 15]
```diff
@@ -1556,6 +1556,71 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
...
+     * @param rNodes array of nodes from which data will be extracted
+     * @param rVar the variable being addressed
+     * @param rData input vector (must be of size rNodes.size()*Dimension)
+     * @param Step database step to which we will write
+     */
+    void SetValuesVector(
+                                ModelPart::NodesContainerType& rNodes,
+                                const Variable<double>& rVar,
+                                const Vector& rData
+                                );
```
Comment (loumalouomega) [inline kratos/utilities/variable_utils.h]:
```suggestion
    void SetValuesVector(
        ModelPart::NodesContainerType& rNodes,
        const Variable<double>& rVar,
        const Vector& rData
        );
```

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Some comments

File: kratos/utilities/variable_utils.cpp (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/utilities/variable_utils.cpp:533]: all the methods in this file have the KRATOS_API tag ...

**Review by loumalouomega — DISMISSED** [review]
I am still not sure about the KRATOS_API, I think it is redundant

**Review by loumalouomega — APPROVED** [review]
https://github.com/KratosMultiphysics/Kratos/blob/542c677df7b329ac7453852eae9b31555b3a3fa5/kratos/utilities/variable_utils.h#L62

See, KRATOS_API is already defined

File: kratos/utilities/variable_utils.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/utilities/variable_utils.cpp:533]: I think it is not required if the class have it, right @roigcarlo ?

Comment (RiccardoRossi) [reply kratos/utilities/variable_utils.cpp:533]: well.. i would prefer not to change it in this specific PR.

however...could it change if the methods are static?

Comment (loumalouomega) [reply kratos/utilities/variable_utils.cpp:533]: Static in a namespace yes, it is the only way. Static in a class already exported is not required

Comment (RiccardoRossi) [reply kratos/utilities/variable_utils.cpp:533]: well, i am not telling you are wrong... but i would keep this as it is, If you like we could try to take out all the KRATOS_API stuff in a different PR

**Pull request merged**